test/hotspot/jtreg/serviceability/jvmti/vthread/FollowReferences/libVThreadStackRefTest.cpp

```diff
@@ -0,0 +1,203 @@
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+#include <jni.h>
+#include <jvmti.h>
+#include <jvmti_common.h>
+#include <atomic>
+#include <string.h>
+
+namespace {
+
+jvmtiEnv *jvmti = nullptr;
+
+const int TAG_START = 100;
+
+struct RefCounters {
+  jint testClassCount;
+  jint *count;
+  jlong *threadId;
+
+  RefCounters(): testClassCount(0), count(nullptr) {}
+
+  void* alloc(JNIEnv* env, jlong size) {
+    unsigned char* ptr;
+    jvmtiError err = jvmti->Allocate(size, &ptr);
+    if (err != JVMTI_ERROR_NONE) {
+      env->FatalError("jvmti->Allocate failed");
+    }
+    memset(ptr, 0, size);
+    return ptr;
+  }
+
+  void init(JNIEnv* env, jint testClassCount) {
+    this->testClassCount = testClassCount;
+    count = (jint*)alloc(env, sizeof(count[0]) *  testClassCount);
+    threadId = (jlong*)alloc(env, sizeof(threadId[0]) *  testClassCount);
+  }
+} refCounters;
+
+}
+
+/////////////////////////////////////////
+// Agent functions
+/////////////////////////////////////////
+jint JNICALL
+HeapReferenceCallback(jvmtiHeapReferenceKind reference_kind,
+                      const jvmtiHeapReferenceInfo* reference_info,
+                      jlong class_tag, jlong referrer_class_tag, jlong size,
+                      jlong* tag_ptr, jlong* referrer_tag_ptr, jint length, void* user_data) {
+  if (class_tag >= TAG_START) {
+    jlong index = class_tag - TAG_START;
+    switch (reference_kind) {
+    case JVMTI_HEAP_REFERENCE_STACK_LOCAL: {
+      jvmtiHeapReferenceInfoStackLocal *stackInfo = (jvmtiHeapReferenceInfoStackLocal *)reference_info;
+      refCounters.count[index]++;
+      refCounters.threadId[index] = stackInfo->thread_id;
+      LOG("Stack local: index = %d, threadId = %d\n",
+          (int)index, (int)stackInfo->thread_id);
+      if (refCounters.count[index] > 1) {
+        LOG("ERROR: count > 1: %d\n", (int)refCounters.count[index]);
+      }
+    }
+    break;
+    case JVMTI_HEAP_REFERENCE_JNI_LOCAL: {
+      jvmtiHeapReferenceInfoJniLocal *jniInfo = (jvmtiHeapReferenceInfoJniLocal *)reference_info;
+      refCounters.count[index]++;
+      refCounters.threadId[index] = jniInfo->thread_id;
+      LOG("JNI local: index = %d, threadId = %d\n",
+          (int)index, (int)jniInfo->thread_id);
+      if (refCounters.count[index] > 1) {
+        LOG("ERROR: count > 1: %d\n", (int)refCounters.count[index]);
+      }
+    }
+    break;
+    default:
+      // unexpected ref.kind
+      LOG("ERROR: unexpected ref_kind for class %d: %d\n",
+          (int)index, (int)reference_kind);
+    }
+  }
+  return JVMTI_VISIT_OBJECTS;
+}
+
+extern "C" JNIEXPORT jint JNICALL
+Agent_OnLoad(JavaVM *vm, char *options, void *reserved) {
+  if (vm->GetEnv(reinterpret_cast<void **>(&jvmti), JVMTI_VERSION) != JNI_OK || jvmti == nullptr) {
```

Comment:
Nit: This line is long and non readable. There are many examples in tests how it is normally done.

Reply:
done